File: projects/relay/engine-java/src/main/java/com/llmeval/relay/engine/Game.java
package com.llmeval.relay.engine;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The turn loop — RELAY's referee. The Java half of ADR-0002.
 *
 * <p>Every ordering decision here is on the conformance path: the order ticks are charged, the
 * order lanes rotate, the order standings sort. A port that is merely <em>correct</em> is not
 * enough — it has to be correct the same way.
 */
public final class Game {

    public static final String ENGINE_VERSION = "0.1.0";

    public static final List<String> COLORS = List.of("red", "green", "yellow", "blue");

    public static final int PHASE_ATTEMPTS = 2;

    public static final int TICK_ANSWER = 2;
    public static final int TICK_ESCALATE = 5;
    public static final int TICK_WRONG = 4;
    public static final int TICK_PASS = 3;

    public static final int ESCALATION_QUOTA = 8;
    public static final int MAX_STALLS = 3;
    public static final int MAX_NOTE_CHARS = 240;

    private final GameConfig config;
    private final EventSink sink;
    private final List<Stage> track;
    private final Map<String, Lane> lanes = new LinkedHashMap<>();
    private final List<NoteRecord> notes = new ArrayList<>();
    private List<Map<String, Object>> turnEvents = new ArrayList<>();

    private int quota;
    private int turn;
    private int rotation = -1;
    private String finisher;

    public Game(GameConfig config, EventSink sink) {
        this.config = config;
        this.sink = sink;
        this.track = Track.generate(new Rng(config.seed), config.stages);
        this.quota = config.escalationQuota;
        for (String color : COLORS) {
            lanes.put(color, new Lane());
        }
    }

    public List<Stage> track() {
        return track;
    }

    public int quota() {
        return quota;
    }

    // -- public ----------------------------------------------------------

    public Outcome play(Map<String, Runner> runners) {
        emitStart(runners);

        List<Map<String, Object>> stages = new ArrayList<>();
        for (Stage stage : track) {
            // Built key by key, not via Map.of: that factory makes an UNORDERED map, and
            // copying one into a LinkedHashMap preserves whatever order it happened to have.
            // The digest sorts keys so conformance would still pass — but the raw transcript
            // would stop being byte-comparable with Python's, which is a guarantee worth more
            // than the two lines it costs.
            Map<String, Object> entry = new LinkedHashMap<>();
            entry.put("id", stage.id());
            entry.put("family", stage.family());
            entry.put("prompt", stage.prompt());
            stages.add(entry);
        }
        emit("track_generated", ordered("stages", stages));

        while (turn < config.maxTurns && finisher == null && !allStalled()) {
            String color = nextLane();
            if (color == null) {
                break;
            }
            turn++;
            playTurn(color, runners.get(color));
        }

        String reason = finisher != null ? "finished" : allStalled() ? "all_stalled" : "turn_cap";

        List<Map<String, Object>> key = new ArrayList<>();
        for (Stage stage : track) {
            Map<String, Object> entry = new LinkedHashMap<>();
            entry.put("id", stage.id());
            entry.put("tier", stage.tier());
            entry.put("answer", stage.answer());
            key.add(entry);
        }

        List<Map<String, Object>> standings = standings();
        Map<String, Object> ended = new LinkedHashMap<>();
        ended.put("reason", reason);
        ended.put("turns_played", turn);
        ended.put("track_key", key);
        ended.put("standings", standings);
        emit("game_ended", ended);
        return new Outcome(reason, turn, standings);
    }

    // -- turn ------------------------------------------------------------

    private void playTurn(String color, Runner runner) {
        turnEvents = new ArrayList<>();
        Lane lane = lanes.get(color);
        Stage stage = track.get(lane.position);

        Map<String, Object> started = new LinkedHashMap<>();
        started.put("player", color);
        started.put("stage", stage.id());
        emit("turn_started", started);

        EscalationDesk desk = desk(color, stage);
        Attempt attempt = ask(color, runner, stage, desk);
        String note = vetNote(color, attempt.note());

        boolean escalated = desk.used();
        int charged = escalated ? TICK_ESCALATE
                : attempt.answer() == null ? TICK_PASS : TICK_ANSWER;
        boolean correct = attempt.answer() != null
                && Track.normalise(attempt.answer()).equals(Track.normalise(stage.answer()));

        String reason;
        if (attempt.answer() == null) {
            lane.passes++;
            lane.stalls++;
            reason = "passed";
        } else if (correct) {
            lane.correct++;
            lane.position++;
            lane.stalls = 0;
            reason = "cleared";
        } else {
            lane.wrong++;
            lane.stalls++;
            charged += TICK_WRONG;
            reason = "missed";
        }

        if (escalated) {
            lane.escalations++;
        }
        lane.ticks += charged;
        lane.history.add(new AttemptRecord(turn, stage.id(), stage.family(), escalated, correct));
        if (note != null) {
            notes.add(new NoteRecord(turn, color, note));
        }

        Map<String, Object> attempted = new LinkedHashMap<>();
        attempted.put("player", color);
        attempted.put("stage", stage.id());
        attempted.put("answer", attempt.answer());
        attempted.put("escalated", escalated);
        attempted.put("correct", correct);
        attempted.put("ticks_charged", charged);
        attempted.put("ticks_total", lane.ticks);
        attempted.put("quota_left", quota);
        attempted.put("note", note);
        emit("stage_attempted", attempted);

        if (lane.position >= track.size()) {
            lane.finished = true;
            finisher = color;
            reason = "finished";
            Map<String, Object> done = new LinkedHashMap<>();
            done.put("player", color);
            done.put("ticks", lane.ticks);
            emit("runner_finished", done);
        }

        Map<String, Object> ended = new LinkedHashMap<>();
        ended.put("player", color);
        ended.put("reason", reason);
        emit("turn_ended", ended);

        if (runner instanceof Reflector reflector) {
            reflector.reflect(new TurnEnd(view(color, stage), color, turn, reason, turnEvents));
        }
    }

    /** The desk is built once per turn, so a runner that asks twice pays twice. */
    private Attempt ask(String color, Runner runner, Stage stage, EscalationDesk desk) {
        for (int attemptNo = 1; attemptNo <= PHASE_ATTEMPTS; attemptNo++) {
            TurnContext ctx = new TurnContext(view(color, stage), color, turn, desk, attemptNo);
            Attempt attempt;
            try {
                attempt = runner.attempt(ctx);
            } catch (RuntimeException e) {  // a broken agent passes; it does not crash the race
                invalid(color, "attempt", "runner error: " + e.getClass().getSimpleName(),
                        attemptNo);
                continue;
            }
            if (attempt == null) {
                invalid(color, "attempt", "no attempt returned", attemptNo);
                continue;
            }
            return attempt;
        }
        return Attempt.pass();
    }

    private EscalationDesk desk(String color, Stage stage) {
        return new EscalationDesk(stage, () -> quota, () -> quota--, config.anchor,
                () -> invalid(color, "escalate", "shared quota exhausted", 1));
    }

    private String vetNote(String color, String note) {
        if (note == null) {
            return null;
        }
        if (note.length() > config.maxNoteChars) {
            invalid(color, "note", "note too long", 1);
            return null;
        }
        return note;
    }

    // -- plumbing --------------------------------------------------------

    private RunnerView view(String color, Stage stage) {
        Lane lane = lanes.get(color);
        List<LaneSnapshot> snapshots = new ArrayList<>();
        for (String c : COLORS) {
            Lane other = lanes.get(c);
            snapshots.add(new LaneSnapshot(c, other.position, other.ticks, other.escalations,
                    other.finished));
        }
        return new RunnerView(color, stage.publicView(), lane.position, lane.ticks, track.size(),
                quota, snapshots, notes, lane.history);
    }

    private String nextLane() {
        for (int i = 0; i < COLORS.size(); i++) {
            rotation = (rotation + 1) % COLORS.size();
            String color = COLORS.get(rotation);
            if (!lanes.get(color).finished) {
                return color;
            }
        }
        return null;
    }

    private boolean stalled(String color) {
        Lane lane = lanes.get(color);
        return lane.finished || lane.stalls > config.maxStalls;
    }

    /** Only a spent quota can strand the table. */
    private boolean allStalled() {
        if (quota > 0) {
            return false;
        }
        for (String color : COLORS) {
            if (!stalled(color)) {
                return false;
            }
        }
        return true;
    }

    private void invalid(String color, String phase, String reason, int attempt) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("player", color);
        payload.put("phase", phase);
        payload.put("reason", reason);
        payload.put("attempt", attempt);
        emit("invalid_action", payload);
    }

    private List<Map<String, Object>> standings() {
        List<String> ranked = new ArrayList<>(COLORS);
        ranked.sort(Comparator
                .comparingInt((String c) -> -lanes.get(c).position)
                .thenComparingInt(c -> lanes.get(c).ticks)
                .thenComparingInt(COLORS::indexOf));

        List<Map<String, Object>> out = new ArrayList<>();
        for (int i = 0; i < ranked.size(); i++) {
            Lane lane = lanes.get(ranked.get(i));
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("player", ranked.get(i));
            row.put("rank", i + 1);
            row.put("stages_cleared", lane.position);
            row.put("ticks", lane.ticks);
            row.put("finished", lane.finished);
            row.put("escalations", lane.escalations);
            row.put("correct", lane.correct);
            row.put("wrong", lane.wrong);
            row.put("passes", lane.passes);
            out.add(row);
        }
        return out;
    }

    // -- emission --------------------------------------------------------

    private static Map<String, Object> ordered(String key, Object value) {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put(key, value);
        return map;
    }

    private void emit(String type, Map<String, Object> payload) {
        Map<String, Object> record = new LinkedHashMap<>();
        record.put("type", type);
        record.put("payload", payload);
        turnEvents.add(record);
        sink.emit(type, payload, turn);
    }

    private void emitStart(Map<String, Runner> runners) {
        List<Map<String, Object>> players = new ArrayList<>();
        for (String color : COLORS) {
            Map<String, Object> meta = new LinkedHashMap<>();
            meta.put("color", color);
            Map<String, Object> configured = config.players.get(color);
            if (configured != null) {
                meta.putAll(configured);
            }
            meta.putIfAbsent("agent",
                    runners.get(color) == null ? "unknown" : runners.get(color).name());
            players.add(meta);
        }

        Map<String, Object> tiers = new LinkedHashMap<>();
        tiers.put("1", 0);
        tiers.put("2", 0);
        tiers.put("3", 0);
        for (Stage stage : track) {
            String k = Integer.toString(stage.tier());
            tiers.put(k, (Integer) tiers.get(k) + 1);
        }

        Map<String, Object> engine = new LinkedHashMap<>();
        engine.put("language", "java");
        engine.put("version", ENGINE_VERSION);

        Map<String, Object> trackShape = new LinkedHashMap<>();
        trackShape.put("stages", track.size());
        trackShape.put("tiers", tiers);

        Map<String, Object> ticks = new LinkedHashMap<>();
        ticks.put("answer", TICK_ANSWER);
        ticks.put("escalate", TICK_ESCALATE);
        ticks.put("wrong", TICK_WRONG);
        ticks.put("pass", TICK_PASS);

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("seed", config.seed);
        payload.put("max_turns", config.maxTurns);
        payload.put("escalation_quota", config.escalationQuota);
        payload.put("max_stalls", config.maxStalls);
        payload.put("max_note_chars", config.maxNoteChars);
        payload.put("ruleset", config.ruleset);
        payload.put("stack", config.stack);
        payload.put("engine", engine);
        payload.put("track", trackShape);
        payload.put("ticks", ticks);
        payload.put("players", players);

        putIfPresent(payload, "profile", config.profile);
        putIfPresent(payload, "prompt_set", config.promptSet);
        putIfPresent(payload, "framework", config.framework);
        putIfPresent(payload, "host", config.host);
        putIfPresent(payload, "anchor", config.anchorMeta);
        emit("game_started", payload);
    }

    private static void putIfPresent(Map<String, Object> payload, String key, Object value) {
        if (value != null) {
            payload.put(key, value);
        }
    }

    private static final class Lane {
        int position;
        int ticks;
        int stalls;
        int escalations;
        int correct;
        int wrong;
        int passes;
        boolean finished;
        final List<AttemptRecord> history = new ArrayList<>();
    }
}

File: projects/relay/engine-python/src/relay_engine/game.py
"""The turn loop — RELAY's referee.

Drives a race to a finisher, a stalled table, or the turn cap, emitting the
shared event stream as it goes. Knows nothing about agents beyond the `Runner`
protocol, and nothing about models at all: the anchor arrives as a plain
callable, the same way the engine has always taken deciders.

Three structural rules a reader should see enforced here rather than trusted:

- **The seal.** Tiers and answers exist in this module and appear in exactly
  one event, `game_ended`. `track_generated` carries prompts only. Nothing a
  runner can reach — its view, its context, its history — has ever held either.
- **The engine performs escalation.** Runners ask the desk; the desk charges
  the shared quota and calls the anchor. `escalated` in the transcript is
  therefore a receipt, not a claim.
- **Ticks are the clock.** Every action's cost is a constant charged here, so
  two runs of one seed produce identical clocks. Real latency is a measurement
  the harness attaches to `llm_call`, and it decides nothing.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .deciders import (
    Attempt, AttemptRecord, Color, COLORS, EscalationDesk, LaneSnapshot,
    NoteRecord, Reflector, Runner, RunnerView, TurnContext, TurnEnd, check,
)
from .events import EventSink
from .rng import Rng
from .track import Stage, TRACK_STAGES, generate

ENGINE_VERSION = "0.1.0"

#: Chances a runner gets to produce a valid attempt before the engine passes
#: for it. Same number as both earlier games.
PHASE_ATTEMPTS = 2

#: The price list. Answering yourself is quick; the anchor is slower even when
#: it is right; being wrong costs on top of whichever you chose.
TICK_ANSWER = 2
TICK_ESCALATE = 5
TICK_WRONG = 4
TICK_PASS = 3

#: Sized by the sweep, not by taste. Above ~12 the pool stops binding at all
#: (quota 12 and quota 20 play identically); at 8 it binds, sharper runners win
#: more, and the weak-runner inversion the bench found is still visible.
ESCALATION_QUOTA = 8
MAX_STALLS = 3
MAX_NOTE_CHARS = 240


@dataclass
class GameConfig:
    seed: int = 1
    max_turns: int = 60
    stages: int = TRACK_STAGES
    escalation_quota: int = ESCALATION_QUOTA
    max_stalls: int = MAX_STALLS
    max_note_chars: int = MAX_NOTE_CHARS
    ruleset: str = "baseline"
    stack: str = "none"
    #: Per-colour metadata for game_started (agent name, model, access).
    players: dict[Color, dict] = field(default_factory=dict)
    #: What escalation consults. None means the perfect anchor of an
    #: engine-only run; a harness passes a closure over its real model, which
    #: receives the PUBLIC stage and never the answer.
    anchor: object = None
    #: Provenance the harness supplies (schema-defined); omitted when None,
    #: which is what keeps conformance vectors byte-stable.
    profile: str | None = None
    prompt_set: dict | None = None
    framework: dict | None = None
    host: dict | None = None
    anchor_meta: dict | None = None


@dataclass
class Outcome:
    reason: str
    turns_played: int
    standings: list[dict]

    @property
    def winner(self) -> Color | None:
        top = self.standings[0]
        return top["player"] if top["finished"] else None


@dataclass
class _LaneState:
    position: int = 0
    ticks: int = 0
    stalls: int = 0
    escalations: int = 0
    correct: int = 0
    wrong: int = 0
    passes: int = 0
    finished: bool = False
    history: list[AttemptRecord] = field(default_factory=list)


class Game:
    def __init__(self, config: GameConfig, sink: EventSink) -> None:
        self.config = config
        self.sink = sink
        self.track: tuple[Stage, ...] = generate(Rng(config.seed), config.stages)
        self.lanes = {c: _LaneState() for c in COLORS}
        self.notes: list[NoteRecord] = []
        self.quota = config.escalation_quota
        self.turn = 0
        self._rotation = -1
        self._finisher: Color | None = None
        self._turn_events: list[dict] = []

    # -- public ----------------------------------------------------------

    def play(self, runners: dict[Color, Runner]) -> Outcome:
        self._emit_start(runners)
        self._emit("track_generated", {
            "stages": [{"id": s.id, "family": s.family, "prompt": s.prompt}
                       for s in self.track],
        })

        while (self.turn < self.config.max_turns and self._finisher is None
               and not self._all_stalled()):
            color = self._next_lane()
            if color is None:
                break
            self.turn += 1
            self._play_turn(color, runners[color])

        if self._finisher is not None:
            reason = "finished"
        elif self._all_stalled():
            reason = "all_stalled"
        else:
            reason = "turn_cap"

        standings = self._standings()
        self._emit("game_ended", {
            "reason": reason,
            "turns_played": self.turn,
            "track_key": [{"id": s.id, "tier": s.tier, "answer": s.answer}
                          for s in self.track],
            "standings": standings,
        })
        return Outcome(reason, self.turn, standings)

    # -- turn ------------------------------------------------------------

    def _play_turn(self, color: Color, runner: Runner) -> None:
        self._turn_events = []
        lane = self.lanes[color]
        stage = self.track[lane.position]
        self._emit("turn_started", {"player": color, "stage": stage.id})

        attempt, desk = self._ask(color, runner, stage)
        note = self._vet_note(color, attempt.note)

        escalated = desk.used
        charged = TICK_ESCALATE if escalated else (
            TICK_PASS if attempt.answer is None else TICK_ANSWER)
        correct = check(attempt.answer, stage)
        if attempt.answer is None:
            lane.passes += 1
            lane.stalls += 1
            reason = "passed"
        elif correct:
            lane.correct += 1
            lane.position += 1
            lane.stalls = 0
            reason = "cleared"
        else:
            lane.wrong += 1
            lane.stalls += 1
            charged += TICK_WRONG
            reason = "missed"

        if escalated:
            lane.escalations += 1
        lane.ticks += charged
        lane.history.append(AttemptRecord(
            self.turn, stage.id, stage.family, escalated, correct))
        if note is not None:
            self.notes.append(NoteRecord(self.turn, color, note))

        self._emit("stage_attempted", {
            "player": color, "stage": stage.id, "answer": attempt.answer,
            "escalated": escalated, "correct": correct,
            "ticks_charged": charged, "ticks_total": lane.ticks,
            "quota_left": self.quota, "note": note,
        })

        if lane.position >= len(self.track):
            lane.finished = True
            self._finisher = color
            reason = "finished"
            self._emit("runner_finished", {"player": color, "ticks": lane.ticks})

        self._emit("turn_ended", {"player": color, "reason": reason})

        if isinstance(runner, Reflector):
            runner.reflect(TurnEnd(
                self._view(color, stage), color, self.turn, reason,
                tuple(self._turn_events),
            ))

    def _ask(self, color: Color, runner: Runner,
             stage: Stage) -> tuple[Attempt, EscalationDesk]:
        """Up to PHASE_ATTEMPTS goes at a valid attempt. The desk is built once
        per turn, so a runner that asks twice pays twice — as it should."""
        desk = self._desk(color, stage)
        for attempt_no in range(1, PHASE_ATTEMPTS + 1):
            ctx = TurnContext(self._view(color, stage), color, self.turn, desk, attempt_no)
            try:
                attempt = runner.attempt(ctx)
            except Exception as exc:  # a broken agent passes; it does not crash the race
                self._invalid(color, "attempt", f"runner error: {type(exc).__name__}", attempt_no)
                continue
            if attempt is None:
                self._invalid(color, "attempt", "no attempt returned", attempt_no)
                continue
            if attempt.answer is not None and not isinstance(attempt.answer, str):
                self._invalid(color, "attempt", "answer must be text", attempt_no)
                continue
            return attempt, desk
        return Attempt(), desk

    def _desk(self, color: Color, stage: Stage) -> EscalationDesk:
        def spend() -> None:
            self.quota -= 1

        def refused() -> None:
            self._invalid(color, "escalate", "shared quota exhausted", 1)

        return EscalationDesk(stage, lambda: self.quota, spend,
                              self.config.anchor, refused)

    def _vet_note(self, color: Color, note: str | None) -> str | None:
        """Notes may lie — that is in-fiction and legal. They may not be
        enormous: an unbounded channel is a cost problem, not a cunning one."""
        if note is None:
            return None
        if not isinstance(note, str):
            self._invalid(color, "note", "note must be text", 1)
            return None
        if len(note) > self.config.max_note_chars:
            self._invalid(color, "note", "note too long", 1)
            return None
        return note

    # -- plumbing --------------------------------------------------------

    def _view(self, color: Color, stage: Stage) -> RunnerView:
        lane = self.lanes[color]
        return RunnerView(
            color, stage.public(), lane.position, lane.ticks, len(self.track),
            self.quota,
            tuple(LaneSnapshot(c, self.lanes[c].position, self.lanes[c].ticks,
                               self.lanes[c].escalations, self.lanes[c].finished)
                  for c in COLORS),
            tuple(self.notes),
            tuple(lane.history),
        )

    def _next_lane(self) -> Color | None:
        for _ in range(len(COLORS)):
            self._rotation = (self._rotation + 1) % len(COLORS)
            color = COLORS[self._rotation]
            if not self.lanes[color].finished:
                return color
        return None

    def _stalled(self, color: Color) -> bool:
        lane = self.lanes[color]
        return lane.finished or lane.stalls > self.config.max_stalls

    def _all_stalled(self) -> bool:
        """Only a spent quota can strand the table: while there is quota left,
        any runner still has a way through its stage."""
        return self.quota <= 0 and all(self._stalled(c) for c in COLORS)

    def _invalid(self, color: Color, phase: str, reason: str, attempt: int) -> None:
        self._emit("invalid_action", {
            "player": color, "phase": phase, "reason": reason, "attempt": attempt,
        })

    def _standings(self) -> list[dict]:
        def key(color: Color):
            lane = self.lanes[color]
            # Further along wins; the clock settles equal progress; colour
            # order settles a dead heat.
            return (-lane.position, lane.ticks, COLORS.index(color))

        ranked = sorted(COLORS, key=key)
        return [
            {
                "player": c,
                "rank": i + 1,
                "stages_cleared": self.lanes[c].position,
                "ticks": self.lanes[c].ticks,
                "finished": self.lanes[c].finished,
                "escalations": self.lanes[c].escalations,
                "correct": self.lanes[c].correct,
                "wrong": self.lanes[c].wrong,
                "passes": self.lanes[c].passes,
            }
            for i, c in enumerate(ranked)
        ]

    # -- emission --------------------------------------------------------

    def _emit(self, type_: str, payload: dict) -> None:
        self._turn_events.append({"type": type_, "payload": payload})
        self.sink.emit(type_, payload, turn=self.turn)

    def _emit_start(self, runners: dict[Color, Runner]) -> None:
        players = []
        for color in COLORS:
            meta = dict(self.config.players.get(color, {}))
            meta.setdefault("agent", getattr(runners.get(color), "name", "unknown"))
            players.append({"color": color, **meta})

        tiers = {"1": 0, "2": 0, "3": 0}
        for stage in self.track:
            tiers[str(stage.tier)] += 1

        payload = {
            "seed": self.config.seed,
            "max_turns": self.config.max_turns,
            "escalation_quota": self.config.escalation_quota,
            "max_stalls": self.config.max_stalls,
            "max_note_chars": self.config.max_note_chars,
            "ruleset": self.config.ruleset,
            "stack": self.config.stack,
            "engine": {"language": "python", "version": ENGINE_VERSION},
            "track": {"stages": len(self.track), "tiers": tiers},
            "ticks": {"answer": TICK_ANSWER, "escalate": TICK_ESCALATE,
                      "wrong": TICK_WRONG, "pass": TICK_PASS},
            "players": players,
        }
        for key, name in (("profile", "profile"), ("prompt_set", "prompt_set"),
                          ("framework", "framework"), ("host", "host"),
                          ("anchor_meta", "anchor")):
            value = getattr(self.config, key)
            if value is not None:
                payload[name] = value
        self._emit("game_started", payload)
